### DISCUSSION TITLE
Removing samples not involved in recombination

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tskit#2703] [Q&A] [answered]
Probably an easy question, but has anyone thought of an algorithm for simplifying a tree sequence to remove only those samples not involved in recombination? Another way to think about this would be to remove samples but retain the same number of trees and breakpoints (that's not *quite* the same, but is a strict requirement of such an algorithm).

I'm unclear if this would require keeping (some) unary nodes in then tree sequence.

**Comment by jeromekelleher** [comment, marked as answer]
You could do something like this:
```python
tree = ts.first()
untouched_by_recomb = []
for sample in ts.samples():
     u = sample
     affected_by_recomb = False
     while u != -1:
         e = ts.edge(tree.edge(u))
         if e.left != 0 or e.right != ts.sequence_length:
              affected_by_recomb = True
              break
         u = tree.parent(u)
    if not affected_by_recomb:
         untouched_by_recomb.append(sample)
```

The idea is that a sample is not involved in recombination if it has a complete tree path all the way back to root, which seems like a reasonable definition, but perhaps its too strong?

I'm sure you could make this implementation more efficient, but this should be a good starting point?

**Reply by hyanwong** [reply]
Thanks @jeromekelleher - this is the sort of thing I was thinking about. I guess we might want to squash edges first too, just to make sure.

**Reply by jeromekelleher** [reply]
simplify should always return a fully squashed ts

**Reply by hyanwong** [reply]
Right, but we want to squash *before* we might have simplified. I guess it's a minor point, though. The routine seems to work well enough for want we need with the covid trees anyway, so 👍

**Reply by jeromekelleher** [reply]
Yes, it does assume that edges are squashed - I think this is true for sc2ts output though
